os/kernel/binary_manager/binary_manager_bootparam.c

```diff
@@ -0,0 +1,152 @@
+ *
+ * Description:
+ *	 This function registers the partition of boot parameters.
+ *
+ ****************************************************************************/
+void binary_manager_register_bppart(int part_num, int part_size)
+{
+	if (part_num < 0 || part_size != BOOTPARAM_PARTSIZE) {
+		bmdbg("ERROR: Invalid part info : num %d, size %d\n", part_num, part_size);
+		return;
+	}
+
+	bp_info.part_num = part_num;
+
+	bmvdbg("[BOOTPARAM] part num %d\n", part_num);
+}
+
+bool is_valid_bootparam(binmgr_bpdata_t *bp_data)
+{
+	if (!bp_data) {
+		bmdbg("Boot param is NULL\n");
+		return false;
+	} else if (bp_data->format_ver == 0 || bp_data->active_idx >= binary_manager_get_kdata()->part_count) {
+		bmdbg("Invalid data. ver: %u, active index: %u, addresses: %x, %x\n", bp_data->version, bp_data->active_idx, bp_data->address[0], bp_data->address[1]);
+		return false;
+	} else if (bp_data->crc_hash != crc32((uint8_t *)bp_data + CHECKSUM_SIZE, sizeof(binmgr_bpdata_t) - CHECKSUM_SIZE)) {
+		bmdbg("Invalid crc32 value, crc32 %u, ver: %u, active index: %u, addresses: %x, %x\n", bp_data->crc_hash, bp_data->version, bp_data->active_idx, bp_data->address[0], bp_data->address[1]);
+		return false;
+	}
+
+	return true;
+}
+
+/*****************************************************************************************************
+ * Name: binary_manager_scan_bootparam
+ *
+ * Description:
+ *	 This function updates the latest boot parameter information by scanning all boot parameters.
+ *
+ *****************************************************************************************************/
+int binary_manager_scan_bootparam(void)
+{
+	int fd;
+	int ret;
+	int bp_idx;
+	int latest_idx = -1;
+	uint32_t latest_ver = 0;
+	binmgr_bpdata_t bp_data[BOOTPARAM_COUNT];
+	char bp_devpath[BINARY_PATH_LEN];
+
+	if (bp_info.part_num < 0) {
+		bmdbg("Invalid Bootparam partition Num : %d\n", bp_info.part_num);
+		return BINMGR_INVALID_PARAM;
+	}
+
+	/* Open dev to access boot parameters */
+	snprintf(bp_devpath, BINARY_PATH_LEN, BINMGR_DEVNAME_FMT, bp_info.part_num);
+	fd = open(bp_devpath, O_RDWR, 0666);
+	if (fd < 0) {
+		bmdbg("ERROR: Get a fd of bootparam, errno %d\n", errno);
+		return BINMGR_OPERATION_FAIL;
+	}
+
+	for (bp_idx = 0; bp_idx < BOOTPARAM_COUNT; bp_idx++) {
+		ret = lseek(fd, BP_SEEK_OFFSET(bp_idx), SEEK_SET);
+		if (ret < 0) {
+			bmdbg("ERROR: Seek Failed, errno %d\n", errno);
+			close(fd);
+			return BINMGR_OPERATION_FAIL;
+		}
+		/* Read bootparam data */
+		ret = read(fd, (FAR uint8_t *)&bp_data[bp_idx], sizeof(binmgr_bpdata_t));
+		if (ret != sizeof(binmgr_bpdata_t)) {
+			continue;
+		} else if (!is_valid_bootparam(&bp_data[bp_idx])) {
+			continue;
+		}
+
+		/* Update the latest version and index */
+		if (latest_ver < bp_data->version) {
+			latest_ver = bp_data->version;
+			latest_idx = bp_idx;
+			bmvdbg("Latest Version %d, index %d\n", latest_ver, latest_idx);
+		}
+	}
+	close(fd);
+
+	if (latest_ver < 0 || latest_idx < 0) {
```

Comment:
`latest_ver` is from `bp_data->version` and `bp_data->version` is unsigned type.
Even `latest_ver` is also unsigned type.
So it cannot be smaller than 0.